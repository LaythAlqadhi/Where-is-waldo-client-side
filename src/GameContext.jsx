import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';

const initialState = {
  state: { loading: false, error: null },
  status: { isStarted: false, isEnded: true },
  message: null,
};

export const GameContext = createContext(initialState);

export function GameProvider({ children }) {
  const [state, setState] = useState({ loading: false, error: null });
  const [status, setStatus] = useState({
    isStarted: false,
    isEnded: false,
    message: null,
  });
  const [message, setMessage] = useState(null);

  return (
    <GameContext.Provider
      value={{
        state,
        setState,
        status,
        setStatus,
        message,
        setMessage,
      }}
    >
      {children}
    </GameContext.Provider>
  );
}

GameProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
