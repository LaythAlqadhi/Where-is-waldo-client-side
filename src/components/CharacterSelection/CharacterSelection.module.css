.menu {
  position: absolute;
  visibility: hidden;
  background-color: var(--secondary-color);
  color: var(--primary-color);
  border-radius: 16px;
  padding: 10px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.5);
}

.menu div {
  position: relative;
  display: flex;
  align-items: center;
  width: 150px;
  transform: scale(1);
  transition: transform .25s;
  cursor: pointer;
}

.menu div::before {
  content: "";
  position: absolute;
  inset: 0;
  opacity: .15;
  border-radius: 16px;
  transition: background-color .25s;
}

.menu div:hover::before {
  background-color: var(--secondary-accent-color);
}

.menu div:active {
  transform: scale(.95);
}

.menu div span {
  align-self: center;
}

.menu div img {
  width: 50px;
  margen: 10px;
}