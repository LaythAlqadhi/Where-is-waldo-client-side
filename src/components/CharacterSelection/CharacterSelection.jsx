import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import castawayTomImgSrc from '../../assets/castaway-tom.jpeg';
import loveyDoveyCoupleImgSrc from '../../assets/lovey-dovey-couple.jpeg';
import thiefImgSrc from '../../assets/thief.jpeg';
import { GameContext } from '../../GameContext';
import style from './CharacterSelection.module.css';

const characterImages = {
  'Castaway Tom': castawayTomImgSrc,
  'Lovey-dovey couple': loveyDoveyCoupleImgSrc,
  Thief: thiefImgSrc,
};

function CharacterSelection({
  id,
  focus,
  setFocus,
  position,
  setCharacters,
  characters,
}) {
  const { state, setState, setStatus, setMessage } = useContext(GameContext);

  const handleSelectCharacter = (charName) => {
    setFocus(!focus);

    const fetchData = async () => {
      const API_URL = 'https://where-is-waldo.adaptable.app/v1';

      try {
        setState({ ...state, loading: true });

        const response = await fetch(`${API_URL}/validate/${id}`, {
          method: 'POST',
          mode: 'cors',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            name: charName,
            coordinates: {
              x: position.x,
              y: position.y,
            },
          }),
        });

        if (response.status >= 400) {
          throw new Error('Server error');
        }

        const result = await response.json();

        if (result.characters) {
          setCharacters(result.characters);

          if (result.characters.length === 0) {
            setStatus({ isStarted: false, isEnded: true });
          }
        }

        setMessage(result.message);
      } catch (err) {
        setState({ ...state, error: err.message });
      } finally {
        setState({ ...state, loading: false });
      }
    };

    fetchData();
  };

  const shouldFlipMenu = position.x !== null && position.x > 50;

  return (
    focus &&
    position.y >= 0 &&
    position.x >= 0 && (
      <div
        className={style.menu}
        style={{
          visibility: 'visible',
          top: `${position.y}%`,
          left: shouldFlipMenu
            ? `calc(${position.x}% - 175px)`
            : `${position.x}%`,
        }}
      >
        {characters &&
          characters.map((char) => (
            <div
              key={char._id}
              onClick={() => handleSelectCharacter(char.name)}
            >
              <img src={characterImages[char.name]} alt="" />
              <span>{char.name}</span>
            </div>
          ))}
      </div>
    )
  );
}

CharacterSelection.propTypes = {
  id: PropTypes.string,
  focus: PropTypes.bool.isRequired,
  setFocus: PropTypes.func.isRequired,
  position: PropTypes.object.isRequired,
  setCharacters: PropTypes.func.isRequired,
  characters: PropTypes.array,
};

CharacterSelection.defaultProps = {
  id: null,
  characters: null,
};

export default CharacterSelection;
