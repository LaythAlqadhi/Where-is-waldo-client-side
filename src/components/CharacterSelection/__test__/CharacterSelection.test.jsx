import React from 'react';
import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';
import { GameContext } from '../../../GameContext';
import CharacterSelection from '../CharacterSelection';

const mockData = {
  id: 'mockId',
  characters: [
    { _id: '1', name: 'Castaway Tom' },
    { _id: '2', name: 'Lovey-dovey couple' },
  ],
  message: 'Mock message',
};

beforeEach(() => {
  global.fetch = vi.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve(mockData),
      status: 200,
    }),
  );
});

it('renders CharacterSelection component correctly with mock data', async () => {
  const setState = vi.fn();
  const setCharacters = vi.fn();
  const setMessage = vi.fn();
  const setFocus = vi.fn();

  render(
    <GameContext.Provider
      value={{
        state: { loading: false, error: null },
        setState,
        status: { isStarted: true, isEnded: false },
        setStatus: vi.fn(),
        setMessage,
      }}
    >
      <CharacterSelection
        id={mockData.id}
        focus={false}
        setFocus={setFocus}
        position={{ x: 0, y: 0 }}
        setCharacters={setCharacters}
        characters={mockData.characters}
      />
    </GameContext.Provider>,
  );

  expect(screen.container).toMatchSnapshot();
});
