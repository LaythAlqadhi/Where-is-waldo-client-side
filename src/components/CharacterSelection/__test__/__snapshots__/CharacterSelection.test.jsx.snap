// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterSelection component renders correctly with mock data 1`] = `undefined`;

exports[`CharacterSelection component renders correctly with mock data 2`] = `undefined`;

exports[`renders CharacterSelection component correctly with mock data 1`] = `undefined`;

exports[`renders CharacterSelection component renders correctly with mock data 1`] = `undefined`;
